### ISSUE TITLE
Can't solve reCaptcha

### ISSUE THREAD

**Issue by Booteille** [dessant/buster#2]
I couldn't make this extension work with reCAPTCHA.

I always get this following message when clicking on the extension icon:

>Your computer or network may be sending automated queries. To protect our users, we can't process your request right now. For more details visit our help page

I tried to reload the webpage without success.

**Comment by dessant** [comment]
Can you load the audio challenge by clicking on the headphone?

**Comment by Booteille** [comment]
Nop, I can't.

**Comment by dessant** [comment]
You're being blocked from accessing the audio challenge, so the extension can't access it either. These false positives can happen, and it may go away if you try a bit later. Unfortunately Google can sometimes also block people with vision impairment this way, further limiting them in their use of the web.

Please contact the website administrator whenever you're blocked from accessing the audio captcha and let them know about your experience. Google also has a forum for giving feedback on their service: https://groups.google.com/forum/#!forum/recaptcha

:disappointed:  https://groups.google.com/forum/#!topic/recaptcha/5TFQ2c4phPU

**Comment by Madis0** [comment]
Maybe that's caused by the way Buster inputs the answer? For example, it could try writing it letter by letter instead of the whole phrase at once.

**Comment by dessant** [comment]
@Madis0, it's possible that it can play a role when the extension is used, but in this case they simply can't access the audio challenge, without the extension being invoved.

I've seen a bunch of users experiencing the same issue:
https://groups.google.com/forum/#!searchin/recaptcha/audio%7Csort:date

**Comment by babyrig** [comment]
I have tried here but just got the message "Something went wrong" using firefox and Opera there is anyway to log/see the errors ? I even create a api to test but in the api dashboard just show only one attempt

**Comment by dessant** [comment]
Hi @babyrig, that is not related to this thread. You can check the logs in Firefox by opening the browser console (Ctrl+Shift+J). Please create a new issue with the results.

**Comment by babyrig** [comment]
@dessant I have the same problem of the OP I understand is the false/positive and google keep us annoying with these recaptcha loop/endless. But with Opera I se have worked two times then later again the  "automated queries" block it's hard because you have made a really briliant job with thes addon the deal now is how to put to work with a blocked recaptcha false/positive, I am trying to found this recipe, after search for a while I saw many complains it's like a stupid thing.

*BTW when I am able to test 'google allow me' the add-on have worked with my api key fine.

**Comment by dessant** [comment]
@babyrig, the "Something went wrong" message should never show up if you are blocked, that's why I wanted to debug that with you in a different issue.

As for getting rid of the block, you could try clearing your browsing data and changing your IP address.

**Comment by DarKWinGTM** [comment]
I found technical to clear you device and get back to whritelist if you get block from google

**Comment by babyrig** [comment]
> I found technical to clear you device and get back to whritelist if you get block from google

there is any mysterious way or you can tell us ?

**Comment by ciampolo** [comment]
Above doesn't work for me I still get automated query even if I change my IP and force https and use a fresh Firefox profile.

I highly suspect that they just check for "biological" input. The new announced recaptcha that google will implement does not require user interaction but just fingerprints your device and (if you pass that device fingerprint) it fingerprints you as a biological entity (ie your behaviour). A Tor browser or Firefox with resistFingerpinting will, even though it can attempt a captcha, always fail it even the old one. The new one will be even less of a chance to solve it (as in literally 0%). 

I highly suspect Google to just check for "biological" input as in, as said before, type letter per letter, random different delays between each letter and even maybe mistyping etc.

Biometrics are something that will always instantly reveal you as a bot if you do not have implemented countermeasures.

**Comment by dessant** [comment]
Randomizing input will help, but there is a much more important aspect: [trusted](https://developer.mozilla.org/en-US/docs/Web/API/Event/isTrusted) events. Extensions cannot fully emulate user input within the browser context, and this somewhat lowers your trust score. This is why the extension works for most people, but may contribute towards triggering the kind of block you can get even without the extension.

The solution is to send user input from outside the browser with the help of a native app and communicate with it through native messaging. This will create events that are indistinguishable from real user input, and may help lower the number of cases where users are blocked. Though again, some users will still be blocked, this is just how reCAPTCHA works.

Until this is implemented, it's possible that you'll have better success by clicking the audio button yourself, and using the extension from the audio page. If you are also blocked from accessing the audio challenge this way, that is not related to the extension. Try clearing your browsing data, changing your IP address, and contact website admins and the reCHAPTCHA team about the false positives you're seeing.

**Comment by ciampolo** [comment]
@dessant I still don't get though why it works fine in most of the cases for me if I use Chrom* browsers while it doesn't when I use Firefox. I even made sure that the Firefox profile is easy to fingerpint as to rule that factor out.
Using Firefox it literally does not work a single time for me regardless of what I do. There are no errors its just the "automated queries".

Otherwise yes I agree fully a native app that simatules biological behaviour would by far be the best.

**Comment by dessant** [comment]
Their algorithms are treating Chrome and Firefox differently, especially when they have different tracking features enabled. More pervasive tracking helps them better identify legitimate traffic, so disabling third-party cookies or using privacy tools could lower your trust score. Maybe Firefox enables less tracking with the default settings.

I've seen posts about Firefox users getting more captchas than those who use Chrome, but this would require controlled testing with many users to verify.

**Comment by dessant** [comment]
@ciampolo, I don't see why would the extension be removed, since it respects the terms of service. It requires user action to perform a single task, and it acts as an accessibility tool that helps humans pass a challenge that they should be able to pass anyway, but they have a hard time because of the design of the reCAPTCHA system.

The other extension you've mentioned sends automated clicks to every ad the browser comes in contact with, and its primary function appears to be to confuse advertisers while also causing financial harm. So these two extensions have not much in common.

**Comment by ciampolo** [comment]
@dessant 
I did some testing of Googles treatment of Chrom* vs Firefox in regards of recaptcha couple months before. This was all done within a fresh profiles and different ips all on a gnu/linux system during multiple months.

Talking about recaptcha Firefox (resistFingerprinting: false) should not get more captchas. The Captchas are, for the most part, about as fast and easy as on Chrome. Though, as before said, after you solved some captchas there was a small (but visible) delay when Firefox loaded new captcha pictures. The more randomly you interacted with the Google object the (ie chaotic movement) the images loaded faster again though it nearly always was visibly slower than on all Chrom* browsers.

Now with resistFingerprinting set to true, as long as you don't have a Tor IP you can try to solve the captchas but will always fail regardless of how correct you are; this is why everyone should always use Firefox with resistFingerprinting enabled.

Also appearently if your size is 1000x900 (again Tor) it becomes even harder to solve the captchas.

The most relevant fact one should keep in mind here is that Chromium and Chrome are made by Google. The exact same people that you try to "fool".

**Comment by ciampolo** [comment]
@dessant 
Well you circumvent Googles captcha and offer that thing inside their own store and even use their own api for it. I highly doubt they like that.  But maybe they won't obviously I cant know.

**Comment by dessant** [comment]
@ciampolo, Buster is an accessibility tool that helps you reach services that you are supposed to have access to as a human. I'm also not trying to "fool" anyone. Let's keep the discussion strictly on topic going forward.

**Comment by baraa272** [comment]
@dessant I want to thank you personally for this masterpiece 
You made millions happy with it 
Thanks from heart

**Comment by baraa272** [comment]
Welcome and Long Live Buster <3

**Comment by anon182739** [comment]
What about solving the image captchas with reverse image search?
https://nakedsecurity.sophos.com/2016/04/08/solving-google-recaptchas-without-using-humans/
https://www.blackhat.com/docs/asia-16/materials/asia-16-Sivakorn-Im-Not-a-Human-Breaking-the-Google-reCAPTCHA-wp.pdf

Also, have the effects of cookies been investigated? If you manage to solve it for Tor IP+no cookies, then it is not possible to block any more.

**Comment by dessant** [comment]
Yeah, I think image recognition will be explored for this extension at some point. I've used speech recognition because it's much easier to implement, and it should have a higher success rate than attempting to solve the visual challenge. I've been told that with VPN the visual challenge is also frequently blocked. I haven't done much tests with cookies, but I guess as you reduce tracking, It'll just block you at some point.

In the short term, we could look into making the visual challenge less painful. For example, the slow fading animations could be removed.

**Comment by jl91569** [comment]
Image recognition might be a bit difficult as Google sometimes adds a huge amount of noise to the pictures.

**Comment by dessant** [comment]
It would only work decently when portions of a single image need to be selected, there are APIs for getting object boundaries within an image, and then the overlapping portions would be selected. The visual challenge with individual images in every rectangle is too low res, and as you say, also grainy sometimes. It would also trigger more retries and blocks because of the lower success rate.

So overall it's not worth focusing on image recognition for now, because our goal in not just to get through, but to do it faster than what it would take for the person who uses the extension to just solve the challenge manually.

**Comment by anon182739** [comment]
The success rate for solving the visual challenge may be low, but the sound challenge is often blocked. It's much rarer to get the visual challenge blocked. For instance, on Tor you can do (difficult) visuals, but no audios.

I tried feeding the images directly into google image, and the success rate was subpar. However, the study I linked seemed to have some success. It might be possible to improve the detection rate by training a neural network only to guess if it's a fire hydrant/store front/vehicle/etc, instead of using a general-purpose one.
There is also the minor caveat that google images is quite block-happy, but it should be possible to replace it.

I uploaded 9 images from a "select all fire hydrants until there are none left".
1. (cars, on highway) - Best guess for this image: freeway
2. (car, parked) - Best guess for this image: international harvester scout [an off-road vehicle produced by International Harvester from 1961 to 1980]
3. (fire hydrant, yellow, on grass) - Best guess for this image: grass
4. (fire hydrant, red, on pavement) - Best guess for this image: horse
5. (fire hydrant, red, on pavement) - Best guess for this image: wood
6. (cars, on highway) - Best guess for this image: intersection
7. (mostly sky) - Best guess for this image: sailboat
8. (cars, on highway) - Best guess for this image: intersection
9. (some people, some bicycles, in front of a house) - Best guess for this image: house

I then repeated it with Wolfram ImageIdentify (same order) - road, car park, water meter, red squirrel, spigot, road, condensation trail, road, person. So it seems it is not useful in an unmodified state.

With Tor, It seems as if the most common is "select all squares", with the second most common being "select until none are left". Sometimes you need multiple correct answers. However, you also have unlimited tries.

I agree in your analysis it should be most fruitful right now to focus on making visuals more friendly and solving audio captchas faster. Have experiments done showing whether introducing a delay changes the ban rate or not? Also, would it be foreign to add an analytics feature to see whether there is a certain type of audio files the solver chokes on, and if so to see if it is possible to come up with some kind of heuristic, such as slowing them down 50% and retrying?

As for user-friendliness, an option could be added to automatically attempt to solve the captcha on open instead of having to open in and then request a solution. Also, it would be much appreciated if it were possible to add such a button to the captcha widget next to the box, which solves it in the background without opening up a bulky captcha window. It is already possible to do this by clicking the box, clicking the yellow button, clicking outside of the window, and waiting. But it would be nice to have it automated.

Just my two cents, cheers and thank you for your good work!

**Comment by dessant** [comment]
@anon182739, thanks for sharing your ideas! I have not yet tested if delaying the input or editing the audio has an effect. I plan to add native app support, and then we could emulate human behavior once we have native input events.

The button was implemented this way for users to generate more native input events, and because solving is not guaranteed. Adding full automation is not planned at this time, though if anyone wants to work on it, I could review a pull request. What's important is to let the user take control of solving if something goes wrong, and to have a choice between manual and automated solving.

**Comment by dessant** [comment]
I've made some tests with the `isTrusted` event property, and it has not made a noticeable difference. The property name does not seem to be present in the reCAPTCHA source, maybe it's not used. So it looks like keyboard input with delays can be emulated without resorting to a native app, though it remains to be seen if it'll help reduce blocking rates.

The 1 and 4 second transitions for image tiles have also been removed separately, which caused consistent rejections, unless I've waited out the time it would have taken for the animation to complete before clicking the new image.

**Comment by dessant** [comment]
Hi @Clarindan! I'm not available for hire at the moment, but you can support the project on [Patreon](https://www.patreon.com/dessant).

**Comment by Clarindan** [comment]
I will support as soon as i have funds available
Sent from Yahoo Mail on Android 
 
  On Thu, Dec 27, 2018 at 8:59 PM, Armin Sebastian<notifications@github.com> wrote:   
Hi @Clarindan! I'm not available for hire at the moment, but you can support the project on Patreon.

**Comment by dessant** [comment]
That's awesome, thanks!

**Comment by Clarindan** [comment]
Can you make a solver that can solve these type of captchas



Sent from Yahoo Mail on Android 
 
  On Thu, Dec 27, 2018 at 9:43 PM, Armin Sebastian<notifications@github.com> wrote:   
That's awesome, thanks!

**Comment by hdpc2015** [comment]
Any solution for this? I tried to open chrome in Incognito, and Busters works without the message of "automated queries".

**Comment by DarKWinGTM** [comment]
Tested in firefox waterfox cyberfox safari chrome chromium and alot browser ETC, When installed this extenions reCapcha will flag automatic block, I think some feature/function/event Google detacted as automatic tools like a bot, If uninstall everything back to normaly

**Comment by Radamanth132** [comment]
Hello,
As every body here i have the same problem "Your computer or network may be sending automated queries. To protect our users, we can't process your request right now. For more details visit our help page", so i used recaptcha solver with paid API and it work, may be it will give you some ideas to solve recaptcha in another location

**Comment by CrendKing** [comment]
My problem is the "Cannot contact reCAPTCHA. Check your connection and try again." message, shown as https://i.imgur.com/SJ42tta.png. You can see the phrase is already solved, but because the extension automatically press enter to complete the challenge, somehow the error message occurs. I tried to modified the extension by commenting out `dispatchEnter(document.querySelector('#recaptcha-verify-button'));` from `solve/main.js` and press the button manually, it works fine. So maybe either remove the automatic press, or add a bit of delay, depending on what kind of detection reCaptcha is employing.

**Comment by wodim** [comment]
This happened to me and no amount of cleaning cookies and switching browsers would help. I also changed my IP address a few times (within the same /16 range, though). It went away after an hour or so.

**Comment by dessant** [comment]
I've created a native companion app for Buster that sends real user input events to the browser, this should make a great difference and reduce blocking considerably. I will likely release it next week, the installation flow still needs to be sorted out.

**Comment by Klapkaak078** [comment]
Dear Desant,

I read that you do not want to let the extension go automatically because you want people to be able to manually bypass it, is it perhaps an option to add an on / off function, Automated bypassing & manual? I know you said that if there are people who would like to see a pull request, I do not think it is necessary to open a pull request.

Cheers Klapkaak078

**Comment by Klapkaak078** [comment]
> This happened to me and no amount of cleaning cookies and switching browsers would help. I also changed my IP address a few times (within the same /16 range, though). It went away after an hour or so.

**Comment by Klapkaak078** [comment]
Dear Desant,

I read that you do not want to let the extension bypass reCAPTCHA automatically because you want that people are able to manually bypass it, is it perhaps an option to add an on / off Switch function, an Automated bypassing & manual? I know you said that if there are people who would like to see somthing similar to open a pull request, I do not think it is necessary to open a pull request for this.

Cheers Klapkaak078

**Comment by dessant** [comment]
> I know you said that if there are people who would like to see somthing similar to open a pull request, I do not think it is necessary to open a pull request for this.

I have a hard time figuring out what you mean by this, could you share your thought process with us?

**Comment by Klapkaak078** [comment]
Sorry for My bad english I'm from the Netherlands, I use google translate, 

What I mean is that many people ask if the extension can be made fully automatic, for Some people it's important to control the captcha manual, And They can choose between the audio solver And the  puzzle challange, on the page where you can change from api you got a on of switch, Maybe you could add a switch there, if They switch it on that the buster extension gonna work full automatic, And if you turn Off you have to click manual in the box And on the icon from the extension. I hope you understand this, i try My best 😅

**Comment by dessant** [comment]
I thought about this in the last few weeks, and I am increasingly against full automation, because it would aid abuse. It is likely that Buster will always require a manual step to request the challenge to be solved.

**Comment by Klapkaak078** [comment]
Dear Dessant, I fully understand you, you even managed to get into the google store which is already a big challenge, and maybe google will not accept this feature after a new reassessment, Keep on going!, my respect!

**Comment by dessant** [comment]
The new option for user input simulation has been released. A client app enables Buster to send native user input to the browser and helps lower the occurrence of difficult challenges and temporary blocks. Follow the instructions to install the app for Windows, Linux or macOS after enabling the feature from the options page.

For now it can be installed for Firefox, I'd like to get some feedback before publishing it for Chrome.

https://addons.mozilla.org/en-US/firefox/addon/buster-captcha-solver/

Please open new issues for any bug you find. I'm especially looking for feedback from macOS users, because I did not have a device to test the client app on macOS.

**Comment by Phrown420** [comment]
Just tested the new version and it still returns a "Your computer or network may be sending automated queries. To protect our users, we can't process your request right now. For more details visit our help page" error message. Waterfox, Windows 10 and the client is installed in the default location

**Comment by dessant** [comment]
@Phrown420, if your trust score is low, you'll still have issues. Try solving a couple of visual captchas as you come across then in the next few days while browsing.

**Comment by Booteille** [comment]
Yup. I can confirm this didn't work for me neither. Because my trust score is probably very low (and can't grow up since I am blocking a lot of things coming from Google. I always have to take nearly 5 minutes to solve the challenge.).

**Comment by as-com** [comment]
I tried this only once on support.dell.com (Firefox Developer Edition on Windows 10) and everything worked after manually clicking the voice challenge button (it moved the mouse to the Buster button and clicked the loading thingy)

**Comment by dessant** [comment]
@Booteille, which Firefox version are you using?

**Comment by Bluscream** [comment]
Can i make buster to execute a autohotkey script?

**Comment by dessant** [comment]
@Bluscream, no, unless you fork the code and add the necessary changes.

**Comment by Bluscream** [comment]
It is possible to execute OS code with WebExtensions? O.o

**Comment by dessant** [comment]
Extensions can communicate with native apps, you can learn more about the feature from the docs: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Native_messaging

**Comment by dessant** [comment]
@Bluscream, for this extension I wrote a small client app in Go because it needed to be cross-platform, but AutoHotkey is also a great choice if you only need to support Windows.

https://github.com/dessant/buster-client/blob/master/cmd/client/main.go

**Comment by Bluscream** [comment]
Yeah i have seen it. Im glad browsers support that nativemessage stuff. I was afraid you would execute command line code with it.

**Comment by Bluscream** [comment]
I was wondering why the delay [here](https://github.com/dessant/buster-client/blob/master/cmd/client/main.go#L63) is set to 800 :thinking: are these milliseconds? So almost a full second. And if yes is the delay after each char or only at the end?

**Comment by dessant** [comment]
Based on the robotgo source it looks like that's characters/minute: https://github.com/go-vgo/robotgo/blob/48a33f8a6d4d34310a225569dbf5784bea653aa6/key/keypress_c.h#L288

Robotgo is a bit buggy btw and the docs are very sparse, for example moving the mouse resonably fast without jumping is not possible on Windows, as soon as the movement delay is increased a bit it becomes extremely slow. I will replace those parts of the code with native API calls at some point if these bugs are not sorted out in robotgo.

**Comment by Booteille** [comment]
@dessant I am on Firefox Developper Edition, on Linux and a lot of user.js configurations to enhance privacy so it's probably related to that.

**Comment by Booteille** [comment]
I just tried to refresh my Firefox on Linux, synchronize with my profile to download same addons as on Windows, and nope, it still doesn't work on Linux.

**Comment by adamwinn** [comment]
I also got "something went wrong, check your browser console".  The Console showed `TypeError: d is undefined`.  Firefox 66.0b8, MacOS

**Comment by dessant** [comment]
Please open new bug reports for issues related to the client app and fill out the issue template.

I'm closing this issue because it has served its purpose, user input simulation has been implemented which should reduce cases where blocking is triggered by the use of the extension.

Please note that you must be able to open the audio challenge for the extension to work, user input simulation will not help if you cannot access the audio challenge.

Clearing your browsing data, changing your IP address, and solving visual challenges as they *naturally* appear during browsing may help increase your trust score enough to be able to access the audio challenge in the future.

Solving more that a couple of challenges a day could cause blocking regardless of the extension.
